### ISSUE TITLE
Upload JavaDocs and Jars for v3.10.0 Release

### ISSUE THREAD

**Issue by twistedpair** [branflake2267/GWT-Maps-V3-Api#77]
Things to complete to finalize release
- Add tags
- Add jars to downloads
- Update Wiki with new build instructions
- Update readme.md
- Upload JavaDocs for v3.9.0 (and v3.8.1)

So, in order, during a release we need to:
1. Take 3.9.0 tag
2. Build
3. Push to AppEngine - ensure all works well
4. Upload to Downloads
5. Update readme.md
6. Update gh-page with JavaDocs (build on Java 7+ for CSS)

**Comment by branflake2267** [comment]
Add Change readme.md docs :) - nevermind, glitched on that, I see it.

**Comment by twistedpair** [comment]
Currently investigating how to automate these processes as I've seen other projects like [GWT-Bootstrap](https://github.com/gwtbootstrap/gwt-bootstrap) do. Looking at using the:
- [github-site-plugin](https://github.com/github/maven-plugins) to deploy javadocs
- [github-download-plugin](https://github.com/github/maven-plugins) to deploy jars as downloads
- [maven-gae-plugin](https://github.com/maven-gae-plugin/maven-gae-plugin) to deploy the showcase

This means the manual step will be to create and push the tag, and update the tag's POM with the proper name. Then just run maven with the `deploy` profile.

Other plugins can also be used to push to Maven Central, but that is out of scope for this issue.

**Comment by branflake2267** [comment]
gae deploy is easiest I think.

**Comment by twistedpair** [comment]
I assume you mean to just use the Eclipse plugin. The only reason to do the Maven bit is so that all of the hoops to go through are automated and don't even need an IDE to be run.

**Comment by branflake2267** [comment]
no maven would deploy it to gae with 'mvn gae:deploy'

**Comment by branflake2267** [comment]
credentials for the gae deploy have to be set in ~/.m2/settings.xml

**Comment by twistedpair** [comment]
Sure. So you can chain goals to phases. That's why the trunk now carries out the `update-file-header` goal when you `mvn compile` not only when you `mvn license:update-file-header`. I was thinking this chaining would be encapsulated in a profile and then something like `mvn deploy -Pdeploy` could run them all. Or they could be run individually. 

No big deal either way. I've been looking at the [Maven Cheatsheet](https://github.com/gwtbootstrap/gwt-bootstrap/wiki/Maven-Cheat-Sheet) for GWT-Bootstrap as a reference for this. Clearly, as you pointout, pgp keys for repo deployment, gae credentials, and Github creds would be setup only in the `settings.xml` file.

**Comment by branflake2267** [comment]
I notice GWTP doing some of this too. I'm tuning it getting ready to build from teamcity/github too. I have some more gwtp 2.5 tests.

**Comment by branflake2267** [comment]
I'll have to add an automated upload to this using the maven github plugin. I changed this slightly for easy releasing.

**Comment by branflake2267** [comment]
I've been working on uploading the javadocs via maven github plugin. So far in GWTP this is what I've done. I'm trying to figure out how to use a doclet called DocLava, which I think is a nicer style, although I can't figure out how to run it in aggregation mode on a multi-module project. Anyway, I wanted to write down config here b/c I may convert back to a doc lava config. This is handy b/c `mvn site` can be run to upload the docs. It overwrites the early config too. So I'm not sure if this will affect having multiple doc versions yet. 

With in the `<plugins/>` mvn site can be run to upload javadocs. I suppose binding the site during the performRelease phase would release with out the snapshot in the config. 

``` xml
<!-- run 'mvn site' to output javadocs to gh-pages -->
<!-- run 'mvn javadoc:javadoc -X' to debug -->
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-javadoc-plugin</artifactId>
    <version>${maven-javadoc-plugin.version}</version>
    <configuration>
        <additionalJOption>-J-Xmx2048m</additionalJOption>
        <failOnError>false</failOnError>
        <links>
            <link>http://download.oracle.com/javase/6/docs/api/</link>
            <link>http://google-web-toolkit.googlecode.com/svn/javadoc/latest/</link>
            <link>http://static.springsource.org/spring/docs/3.0.x/javadoc-api/</link>
            <link>http://google-gin.googlecode.com/svn/trunk/javadoc/</link>
            <link>http://google-guice.googlecode.com/svn/trunk/javadoc/</link>
            <link>http://aopalliance.sourceforge.net/doc/</link>
            <link>https://developers.google.com/appengine/docs/java/javadoc/</link>
            <link>http://docs.guava-libraries.googlecode.com/git/javadoc/</link>
        </links>
        <outputDirectory>${project.build.directory}/javadoc</outputDirectory>
        <reportOutputDirectory>${project.reporting.outputDirectory}/javadoc</reportOutputDirectory>
        <excludes>
            <exclude>*.txt</exclude>
            <exclude>**/*.txt</exclude>
        </excludes>
        <header>
            <![CDATA[
                <a href="https://github.com/ArcBees/GWTP" target="_blank">Back to GWTP Home</a>
            ]]>
        </header>
    </configuration>
    <executions>
        <execution>
            <id>attach-javadocs</id>
            <phase>site</phase>
            <goals>
                <goal>aggregate</goal>
            </goals>
        </execution>
        <!-- <execution>
            <id>attach-javadocs2</id>
            <phase>performRelease</phase>
            <goals>
                <goal>aggregate</goal>
            </goals>
        </execution> -->
    </executions>
</plugin>

<!-- run 'mvn ghSite:site -X' to debug -->
<plugin>
    <groupId>com.github.github</groupId>
    <artifactId>site-maven-plugin</artifactId>
    <version>${github.version}</version>
    <configuration>
        <!-- <dryRun>true</dryRun> -->
        <outputDirectory>javadoc</outputDirectory>
        <path>${project.version}</path>
        <message>Creating site for ${project.version}</message>
        <!-- sets the ~/.m2/setting.xml profile -->
        <server>github</server>
        <outputDirectory>${project.build.directory}/site/javadoc</outputDirectory>
    </configuration>
    <executions>
        <execution>
            <goals>
                <goal>site</goal>
            </goals>
            <phase>site</phase>
        </execution>
    </executions>
</plugin>
```

**Comment by branflake2267** [comment]
What do you think of the DocLava style?
http://google-guice.googlecode.com/git/javadoc/packages.html

**Comment by twistedpair** [comment]
@branflake2267 that or the 1.7 JDK look fine to me.

**Comment by branflake2267** [comment]
Yeah, 1.7 is ok with me to. :) I'm trying doc lava b/c it's been requested for GWTP. :)

**Comment by twistedpair** [comment]
Totally indifferent.  Anything is better than javadoc classic.

**Comment by branflake2267** [comment]
I'm sticking with 1.7 b/c it's far easier. I messed around with DocLava style and what a pain!!

**Comment by branflake2267** [comment]
By the way, great job on that. :)

**Comment by branflake2267** [comment]
Have you ever worked with building multiple javadocs version in the maven setup?

**Comment by branflake2267** [comment]
I've got GWTP uploading the docs, but I gave up on building past versions of the javadocs for the multi-module. I have a feeling that if another module was setup it could gather all the versioned jars up and then build the java docs from the jars, but I haven't tried it yet. This might be handy for automating the entire process of future and past javadocs. 

https://github.com/ArcBees/GWTP/blob/master/pom.xml#L483 - javadocs build
https://github.com/ArcBees/GWTP/blob/master/pom.xml#L516 - gh-pages deploy
